src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentHelper.cs

```diff
@@ -30,92 +33,129 @@ public void Contextualize([NotNull] ViewContext viewContext)
+        public void RenderInvoke([NotNull] string name, params object[] arguments)
         {
-            var componentType = SelectComponent(name);
-            InvokeCore(_viewContext.Writer, componentType, args);
+            var descriptor = SelectComponent(name);
+            InvokeCore(_viewContext.Writer, descriptor, arguments);
         }
 
-        public void RenderInvoke([NotNull] Type componentType, params object[] args)
+        public void RenderInvoke([NotNull] Type componentType, params object[] arguments)
         {
-            InvokeCore(_viewContext.Writer, componentType, args);
+            var descriptor = SelectComponent(componentType);
+            InvokeCore(_viewContext.Writer, descriptor, arguments);
         }
 
-        public async Task<HtmlString> InvokeAsync([NotNull] string name, params object[] args)
+        public async Task<HtmlString> InvokeAsync([NotNull] string name, params object[] arguments)
         {
-            var componentType = SelectComponent(name);
-            return await InvokeAsync(componentType, args);
+            var descriptor = SelectComponent(name);
+
+            using (var writer = new StringWriter())
+            {
+                await InvokeCoreAsync(writer, descriptor, arguments);
+                return new HtmlString(writer.ToString());
+            }
         }
 
-        public async Task<HtmlString> InvokeAsync([NotNull] Type componentType, params object[] args)
+        public async Task<HtmlString> InvokeAsync([NotNull] Type componentType, params object[] arguments)
         {
+            var descriptor = SelectComponent(componentType);
+
             using (var writer = new StringWriter())
             {
-                await InvokeCoreAsync(writer, componentType, args);
+                await InvokeCoreAsync(writer, descriptor, arguments);
                 return new HtmlString(writer.ToString());
             }
         }
 
-        public async Task RenderInvokeAsync([NotNull] string name, params object[] args)
+        public async Task RenderInvokeAsync([NotNull] string name, params object[] arguments)
         {
-            var componentType = SelectComponent(name);
-            await InvokeCoreAsync(_viewContext.Writer, componentType, args);
+            var descriptor = SelectComponent(name);
+            await InvokeCoreAsync(_viewContext.Writer, descriptor, arguments);
         }
 
-        public async Task RenderInvokeAsync([NotNull] Type componentType, params object[] args)
+        public async Task RenderInvokeAsync([NotNull] Type componentType, params object[] arguments)
         {
-            await InvokeCoreAsync(_viewContext.Writer, componentType, args);
+            var descriptor = SelectComponent(componentType);
+            await InvokeCoreAsync(_viewContext.Writer, descriptor, arguments);
         }
 
-        private Type SelectComponent([NotNull] string name)
+        private ViewComponentDescriptor SelectComponent([NotNull] string name)
```

Comment:
>Why SelectComponent then?

https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Core/IActionSelector.cs#L11

>I ran into a similar situation when designing TagHelpers and their corresponding TagHelperDescriptors. Became very confusing very fast when the name "Descriptor" was intermingled with "TagHelper".
> One good example is the TagHelperDescriptorProvider used to have a GetTagHelper method which returned TagHelperDescriptors. On multiple occasions it lead to unnecessary confusion when explaining how TagHelpers work, confusion is never bueno.

This code doesn't have similar requirements. The selector is used in one place, in a piece of code that never touches the `ViewComponent` instance.

In my mind adding `Descriptor` all over this code doesn't add any clarity, it's obvious that's what it does. The logical operation taking place here is selecting *which view component to use*. The fact that it returns a descriptor is just a detail of how/what.

In the one place where you **do** care about distinguishing the VC from the VCD they are already labeled as such (`ViewComponentContext`)